<!-- -->
<launch>
  <!-- Visualiztion -->
  <arg name="viz" default="true"/>

  <!-- Which modules to start -->
  <arg name='sensor_fusion' default="false"/>
  <arg name='localization' default="false"/>

  <!-- CARLA connection -->
  <arg name='host' default='localhost'/>
  <arg name='port' default='2000'/>
  <arg name='timeout' default='10'/>

  <!-- Ego vehicle -->
  <arg name='role_name' default='ego_vehicle'/>
  <arg name="vehicle_filter" default='vehicle.audi.etron'/>
  <arg name="sensor_definition_file" default="$(find mike_av_stack)/scripts/sensor_fusion/configs/sensors.json"/>
  <arg name="spawn_point" default="65.516594,7.808423,0.275307,0.0,0.0,0.0"/>
  <!-- use comma separated format "x,y,z,roll,pitch,yaw" -->

  <!-- Map to load on startup (either a predefined CARLA town (e.g. 'Town01'), or a OpenDRIVE map file) -->
  <arg name="town" default='Town03'/>

  <!-- Synchronous mode-->
  <arg name='synchronous_mode' default=''/>
  <arg name='synchronous_mode_wait_for_vehicle_control_command' default=''/>
  <arg name='fixed_delta_seconds' default='0.008333333'/>

  <!-- Localization settings -->
  <arg name="map_name" default="map.pcd"/>
  <!-- Options: icp, ndt, icps -->
  <arg name="scan_matching_algorithm" default="icp"/>
  <arg name="iters" default="100"/>


  <include file="$(find carla_ros_bridge)/launch/carla_ros_bridge.launch">
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='town' value='$(arg town)'/>
    <arg name='timeout' value='$(arg timeout)'/>
    <arg name='synchronous_mode' value='$(arg synchronous_mode)'/>
    <arg name='synchronous_mode_wait_for_vehicle_control_command' value='$(arg synchronous_mode_wait_for_vehicle_control_command)'/>
    <arg name='fixed_delta_seconds' value='$(arg fixed_delta_seconds)'/>
  </include>

  <include file="$(find carla_ego_vehicle)/launch/carla_example_ego_vehicle.launch">
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='timeout' value='$(arg timeout)'/>
    <arg name='role_name' value='$(arg role_name)'/>
    <arg name="vehicle_filter" value='$(arg vehicle_filter)'/>
    <arg name="sensor_definition_file" value='$(arg sensor_definition_file)'/>
    <arg name="spawn_point" value="$(arg spawn_point)"/>
  </include>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find carla_ros_bridge)/config/carla_default_rviz.cfg.rviz"/>

  <include file="$(find carla_manual_control)/launch/carla_manual_control.launch">
    <arg name='role_name' value='$(arg role_name)'/>
  </include>

  <!-- <node pkg="carla_spawn_objects" type="carla_spawn_objects" name="carla_spawn_objects">
    <arg name="spawn_point"
  </node> -->

  <!-- My point cloud stacker -->
  <include file="$(find mike_av_stack)/launch/tools.launch"/>
  <!-- <include file="$(find mike_av_stack)/launch/traffic_manager.launch"/> -->

  <!-- Sensor fusion node -->
  <group if="$(arg sensor_fusion)">
    <include file="$(find mike_av_stack)/launch/sensor_fusion.launch">
      <arg name="viz" value='$(arg viz)'/>
    </include>
  </group>

  <!-- Localization node -->
  <group if="$(arg localization)">
    <include file="$(find mike_av_stack)/launch/localization.launch">
      <arg name="map_name" value='$(arg map_name)'/>
      <!-- Options: icp, ndt, icps -->
      <arg name="scan_matching_algorithm" value='$(arg scan_matching_algorithm)'/>
      <arg name="viz" value='$(arg viz)'/>
      <arg name="iters" value='$(arg iters)'/>
    </include>
  </group>

</launch>